src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs

```diff
@@ -676,37 +679,69 @@ public int ReadSequenceLength()
 
         private int DeserializeLength(int lengthOfLength)
         {
-            int result;
-            if (PeekByte() == 0)
+            if (lengthOfLength == 0 || (uint)lengthOfLength > 4)
             {
-                throw new RlpException("Length starts with 0");
+                ThrowArgumentOutOfRangeException(lengthOfLength);
```

Comment:
QQ: How does this optimize the code? Or its just a styling preference?

Reply:
Is a bit of a lower level optimisation based on both how the Jit operates and the CPU:

1. The Jit complies in whole method chunks
    (a). Time taken to Jit is based on per method size (think of a method as the indivisible block of code)
    (b). The Jit clears stack for all locals in the method; whether or not they are used in that current run though
    (c). Smaller methods are more likely considered for inlining (the biggest of optimizations)
    (d). The Jit has an optimization for "Does not return" methods; i.e. ones that only throw as it knows it doesn't need to clean up the stack and prepare the return buffer etc like it would for a normal method call.
2. A CPU core only has 32kB of L1 Instruction cache

So what I'm doing is a hot/cold code split; moving the cold code that never really gets called out of the method so other optimisations can take place. 

Looking at it for this example
----

1. (b) can be seen in the method preamble asm; with the exceptions directly in the code (red) extra local variables are needed for the string interpolation. Removing them (green) takes out 3x 16 byte stack clears and adds one additional 8 bit stack clear (40 bytes less stack cleared, or less memory writes)

```diff
<        sub      rsp, 96
>        sub      rsp, 48
>        mov      qword ptr [rsp+20H], rax       ; additional 8 byte stack clear
         mov      qword ptr [rsp+28H], rax
<        vxorps   xmm4, xmm4
<        vmovdqa  xmmword ptr [rsp+30H], xmm4    ; clearing 16 bytes of stack
<        vmovdqa  xmmword ptr [rsp+40H], xmm4    ; clearing 16 bytes of stack
<        vmovdqa  xmmword ptr [rsp+50H], xmm4    ; clearing 16 bytes of stack
```
----
1 (a), (c) are shown in what is removed from the method by this change. All the code to do string interpolation for the message, newing up an exception and throwing it is all taken out, replaced by a very minimal method call + cpu fault `int3` [1. (d)].
```diff
<        lea      rcx, [rsp+38H]
<        mov      edx, 27
<        mov      r8d, 1
<        call     [System.Runtime.CompilerServices.DefaultInterpolatedStringHandler:.ctor(int,int):this]
<        mov      ecx, 0x55C
<        mov      rdx, 0x7FFF7023AD00
<        call     CORINFO_HELP_STRCNS
<        mov      rdx, rax
<        lea      rcx, [rsp+38H]
<        call     [System.Runtime.CompilerServices.DefaultInterpolatedStringHandler:AppendStringDirect(System.String):this]
<        lea      rcx, [rsp+38H]
<        mov      edx, esi
<        call     [System.Runtime.CompilerServices.DefaultInterpolatedStringHandler:AppendFormatted[int](int):this]
<        mov      rcx, 0x7FFF7023AC90
<        call     CORINFO_HELP_NEWSFAST
<        mov      rsi, rax
<        lea      rcx, [rsp+38H]
<        call     [System.Runtime.CompilerServices.DefaultInterpolatedStringHandler:ToStringAndClear():System.String:this]
<        mov      rdx, rax
<        mov      rcx, rsi
<        call     [System.InvalidOperationException:.ctor(System.String):this]
<        mov      rcx, rsi
<        call     CORINFO_HELP_THROW
>        mov      ecx, esi
>        call     [Nethermind.Serialization.Rlp.RlpStream:<DeserializeLength>g__ThrowArgumentOutOfRangeException|65_1(int)]
>        int3     
...
<        mov      rcx, 0x7FFF70323338
<        call     CORINFO_HELP_NEWSFAST
<        mov      rsi, rax
<        mov      rcx, rsi
<        call     [System.Exception:.ctor():this]
<        mov      ecx, 0x594
<        mov      rdx, 0x7FFF7023AD00
<        call     CORINFO_HELP_STRCNS
<        lea      rcx, bword ptr [rsi+10H]
<        mov      rdx, rax
<        call     CORINFO_HELP_ASSIGN_REF
<        mov      rcx, rsi
<        call     CORINFO_HELP_THROW
>        call     [Nethermind.Serialization.Rlp.RlpStream:<DeserializeLength>g__ThrowInvalidData|65_0()]
>        int3     
```
This is both faster to Jit compile (it doesn't compile the throw methods until they are called, when the Jit time will be minimal compared to the cost of throwing the exception anyway)

---

Which then leads to 2. The machine code has been reduced from 323 bytes of a mix of hot and cold code, to 132 bytes of all hot code. As the CPU only have 32kB of L1 instruction cache (per core); this reduced size means its more likely to stay in the L1 cache for longer and not need to be reloaded from L2, L3 or main memory, so run faster.

```diff
< ; Total bytes of code: 323
> ; Total bytes of code: 132
```

Hope that helps?

Reply:
Oh and the smaller size means its more likely to be considered for inlining; whereas if the inline needs to bring in all the exception construction etc, then its less likely to be considered as it will bloat the method it inlines to

Reply:
At a more extreme form, check out the [ThrowHelper.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs) class in the runtime 😅

Reply:
Very nice. Thank you.